File: src/com/themaskedcrusader/tmcz/modules/bleed/BleedListener.java
/*
 * Copyright 2013 Topher Donovan (themaskedcrusader.com)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.themaskedcrusader.tmcz.modules.bleed;

import com.themaskedcrusader.bukkit.Random;
import com.themaskedcrusader.bukkit.config.Messages;
import com.themaskedcrusader.bukkit.config.Settings;
import com.themaskedcrusader.tmcz.data.PlayerUtil;
import com.themaskedcrusader.tmcz.data.Status;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

public class BleedListener implements Listener {
    Random random = new Random();

    public BleedListener(JavaPlugin plugin) {
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void bleedPlayer(EntityDamageEvent event) {
        if (Bleed.isAllowed(event.getEntity())) {
            Player player = (Player) event.getEntity();
            int chance = Settings.getConfig().getInt(Bleed.CHANCE);
            if (chance > random.getChance() && !PlayerUtil.isStatusSet(player, Status.BLEEDING)) {

                if (
                        event.getCause() == EntityDamageEvent.DamageCause.ENTITY_ATTACK ||
                        event.getCause() == EntityDamageEvent.DamageCause.ENTITY_EXPLOSION ||
                        event.getCause() == EntityDamageEvent.DamageCause.FALL ||
                        event.getCause() == EntityDamageEvent.DamageCause.FIRE_TICK ||
                        event.getCause() == EntityDamageEvent.DamageCause.PROJECTILE
                   ) {

                    int seconds = Settings.getConfig().getInt(Bleed.SECONDS) * 20;
                    PotionEffect bleed = new PotionEffect(PotionEffectType.getById(15), seconds, 1);
                    player.sendMessage(ChatColor.RED + Messages.getConfig().getString(Bleed.HIT_MSG));
                    player.addPotionEffect(bleed);
                    PlayerUtil.setStatus(player, Status.BLEEDING, true);
                }
            }
        }
    }

    @EventHandler
    public void stopBleedWithBandage(PlayerInteractEvent event) {
        if (Bleed.isAllowed(event.getPlayer())) {
            Player player = event.getPlayer();
            if ((event.getAction() == Action.RIGHT_CLICK_AIR || event.getAction() == Action.RIGHT_CLICK_BLOCK) &&
                    player.getItemInHand().getType() == Material.PAPER) {
                useBandage(player);
            }
        }
    }

    private void useBandage(Player player) {
        if (PlayerUtil.isStatusSet(player, Status.BLEEDING)) {
            PlayerUtil.setStatus(player, Status.BLEEDING, false);

            PotionEffect regen = new PotionEffect(PotionEffectType.REGENERATION, 4, 2);
            player.addPotionEffect(regen);

            if (player.getItemInHand().getAmount() > 1) {
                player.getItemInHand().setAmount(player.getItemInHand().getAmount() - 1);
            } else {
                player.setItemInHand(null);
            }

            player.sendMessage(ChatColor.GREEN + Messages.getConfig().getString(Bleed.STOP_1_MSG));
        }
    }
}

File: src/main/kotlin/me/schalk/spigot/tmcz/modules/bleed/BleedListener.kt
/*
 * Copyright (c) 2013 - 2023 Christopher Schalk
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package me.schalk.spigot.tmcz.modules.bleed

import me.schalk.spigot.lib.math.getChancePercentage
import me.schalk.spigot.tmcz.data.GameData
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.Particle
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.block.Action
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.PlayerInteractEntityEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

class BleedListener(val plugin: JavaPlugin) : BleedModule() {

    private val bandageRegen = PotionEffect(PotionEffectType.REGENERATION, 80, 2, false, false, false)
    private val healerReward = PotionEffect(PotionEffectType.ABSORPTION, 400, 0, false, false, false)

    private val bleedingCauses = listOf(
        EntityDamageEvent.DamageCause.ENTITY_ATTACK,
        EntityDamageEvent.DamageCause.ENTITY_EXPLOSION,
        EntityDamageEvent.DamageCause.FALL,
        EntityDamageEvent.DamageCause.FIRE_TICK,
        EntityDamageEvent.DamageCause.PROJECTILE,
        EntityDamageEvent.DamageCause.CONTACT,
    )

    init {
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun bleedPlayer(event: EntityDamageEvent) {
        if (isAllowed(event.entity)) {
            val player = event.entity as Player
            if (!GameData.getPlayer(event.entity as Player).bleeding && getChancePercentage() >= CHANCE_TO_BLEED) {
                if (event.cause in bleedingCauses) {
                    player.sendMessage(ChatColor.RED.toString() + HIT_MSG)
                    player.spawnParticle(Particle.DAMAGE_INDICATOR, player.location, PARTICLES, 0.2, 0.2, 0.2, 0.01)
                    GameData.getPlayer(event.entity as Player).bleeding = true
                }
            }
        }
    }

    @EventHandler(priority = EventPriority.HIGH)
    fun stopBleedingUponDeath(event: PlayerDeathEvent) {
        if (isAllowed(event.entity)) {
            if (GameData.getPlayer(event.entity).bleeding && event.entity.lastDamageCause?.cause == EntityDamageEvent.DamageCause.CUSTOM)
                event.deathMessage = DEATH_MESSAGE?.replace("__player", event.entity.displayName)
            GameData.getPlayer(event.entity).bleeding = false
        }
    }

    @EventHandler
    fun selfStopBleedingWithBandage(event: PlayerInteractEvent) {
        if (isAllowed(event.player)) {
            val player = event.player
            if ((event.action == Action.RIGHT_CLICK_AIR || event.action == Action.RIGHT_CLICK_BLOCK) &&
                player.inventory.itemInMainHand.type == BANDAGE
            ) {
                useBandage(player)
            }
        }
    }

    @EventHandler
    fun healerStopsBleedingWithBandage(event: PlayerInteractEntityEvent) {
        if (isAllowed(event.player)) {
            val healer = event.player
            val healerName = healer.name
            if (event.rightClicked.type == EntityType.PLAYER && healer.inventory.itemInMainHand.type == BANDAGE) {
                val healedName = event.rightClicked.name
                val healed = Bukkit.getPlayerExact(healedName)
                if (healed != null) {
                    if (GameData.getPlayer(healed).bleeding) {
                        GameData.getPlayer(healed).bleeding = false
                        healed.sendMessage(ChatColor.GREEN.toString() + HEALED_MESSAGE?.replace("__healer", healerName))
                        healed.addPotionEffect(bandageRegen)

                        healer.inventory.itemInMainHand.amount--
                        healer.sendMessage(ChatColor.GREEN.toString() + HEALER_MESSAGE?.replace("__healed", healedName))
                        healer.addPotionEffect(healerReward)
                        healer.addPotionEffect(bandageRegen)
                        GameData.getPlayer(healer).addPlayerHeals()
                    }
                }
            }
        }
    }

    private fun useBandage(player: Player) {
        if (GameData.getPlayer(player).bleeding) {
            GameData.getPlayer(player).bleeding = false
            player.addPotionEffect(bandageRegen)
            if (player.inventory.itemInMainHand.amount > 1) {
                player.inventory.itemInMainHand.amount--
            } else {
                player.inventory.itemInMainHand.type = Material.AIR
            }
            player.sendMessage(ChatColor.GREEN.toString() + SELF_STOP)
        }
    }
}